### PR TITLE
Pump pressure calculation now respects temperature.

### PR DESCRIPTION
**Pull request by vincentiusvin**
## About The Pull Request
Ever pump hot nob into a tank of oxygen, see the pressure rise to 8000 kpa for a few seconds and have that deafening explosion sound play right after? This attempts to eliminate those by making the pressure transfer proc thingy in gas_mixture account for resulting temperature. The old one is preserved if the temperature difference isn't too big because it's most definitely more performant. The minimum temperature difference of 5 is completely arbitrary and I won't mind changing if asked. Math is a bit messy by necessity but I hope the comment is adequate.

~~While I'm at it i also changed the return value of those procs to return a removed gasmix instead of directly merging. I sort of needed this for a clean implementation of assume_air in #62284. Will need more testing though.~~ Found a better solution.

Writes a small unit test to check for this too.

image
Look at that sweet margin. Old one for ref:
image

https://www.desmos.com/calculator/gq9ozrgujn

## Why It's Good For The Game
Cuts down on one unintended cause of explosions.

## Changelog
:cl:
fix: Pumping hot gas into cold gas using the regular pumps should no longer explode.
code: Some math changes to how pressure based pumps calculate pressure (both portable and fixed). Report if problems are found.
/:cl:

### CODE REVIEW

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Test ✔

**Review by Mothblocks — APPROVED** [review]
Sorry I meant

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
Like the idea, implementation is a mite messy

**Comment by vincentiusvin** [comment]
Closed this due to problems with the magnitude of numbers involved, seems like it can be circumvented somewhat easily using Newton's method. Problem will now only arise when `thermal_energy()` is roughly around (give or take one or two) the same magnitude as the e38 limit as opposed to sqrt(e38) when using the quadratic equation.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 1]
```diff
@@ -513,70 +513,90 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
...
+	*/
+	var/ntvalue = target_pressure * output_air.volume / R_IDEAL_GAS_EQUATION
+	var/a_value = thermal_energy() / total_moles()
+	var/b_value = (output_air.total_moles() * thermal_energy() / total_moles())  + output_air.thermal_energy() - (ntvalue*heat_capacity()/total_moles())
+	var/c_value = (-1 * ntvalue * output_air.heat_capacity() + (output_air.total_moles()) * output_air.thermal_energy())
+
+	var/square_rooted = (b_value ** 2) - (4 * a_value * c_value)
+	if(square_rooted > 0)
+		var/result_one = ((-1 * b_value) + sqrt(square_rooted)) / (2 * a_value)
+		var/result_two = ((-1 * b_value) - sqrt(square_rooted)) / (2 * a_value)
```
[the remark this answers is not in the document]
Comment (vincentiusvin) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: This question still stands, its a hot jumble but I'm really unsure on how to tidy it.

**Comment by LemonInTheDark** [comment]
what's our rough limits before breakdown then? x plasma at x temperature?
I realize it's a sliding scale, but I want at least some idea of the limits of sanity you're setting here.

Also do we have a way to detect failure? or will it lead to negative temperatures and such

**Comment by vincentiusvin** [comment]
> Also do we have a way to detect failure? or will it lead to negative temperatures and such
Yep! lower_limit and upper_limit in the gas_pressure_calculate proc are narrow estimates for how big or small the numbers should be.
Ill get you more data in a moment.

**Comment by vincentiusvin** [comment]
Method is same as the unit test i wrote, with a for loop assigning 1e2, 1e3, 1e4, etc till 1e26 to a var/i applied to:
```
for (var/i in list(1e2, 1e3, 1e4, 1e5, 1e6, 1e7, 1e8, 1e9, 1e10, 1e11, 1e12, 1e13, 1e14, 1e15, 1e16, 1e18, 1e19, 1e20, 1e21, 1e22, 1e23, 1e24, 1e25, 1e26))

first_mix.gases[/datum/gas/hypernoblium][MOLES] = i
second_mix.gases[/datum/gas/tritium][MOLES] = 200
first_mix.temperature = T20C + i
second_mix.temperature = T20C
```

<summary> Actual pressure: </summary>
<details>
Gas pressure pumping test finished. i=100 Expected pressure = 2936.08 +/- 5. Got 2936.08. <br/>
Gas pressure pumping test finished. i=1000 Expected pressure = 2936.08 +/- 5. Got 2936.13.  <br/>
Gas pressure pumping test finished. i=10000 Expected pressure = 2936.08 +/- 5. Got 2931.79. <br/>
Gas pressure pumping test finished. i=100000 Expected pressure = 2936.08 +/- 5. Got 2934.39. <br/>
Gas pressure pumping test finished. i=1000000 Expected pressure = 2936.08 +/- 5. Got 3267. <br/>
Gas pressure pumping test finished. i=10000000 Expected pressure = 2936.08 +/- 5. Got 2436.08. <br/>
Gas pressure pumping test finished. i=1e+009 Expected pressure = 2936.08 +/- 5. Got 2436.08. <br/>
Gas pressure pumping test finished. i=1e+010 Expected pressure = 2936.08 +/- 5. Got 2436.08. <br/>
etc..
</details>

starting at i=1e8
Runtime in stack_trace.dm,3: Failed to solve pressure quadratic equation. A: 2.00001e+011. B: 4e+013. C:-2.40674e+007. Current value = 0
Runtime in stack_trace.dm,3: Failed to solve pressure quadratic equation. A: 2e+012. B: 4e+014. C:-2.40674e+007. Current value = 0
Runtime in stack_trace.dm,3: Failed to solve pressure quadratic equation. A: 2e+013. B: 4e+015. C:-2.40674e+007. Current value = 0
Runtime in stack_trace.dm,3: Failed to solve pressure quadratic equation. A: 2e+014. B: 4e+016. C:-2.40674e+007. Current value = 0
etc...


Starting at i=e15
Runtime in stack_trace.dm,3: Newton's Approximation for pressure failed after 10 iterations. Current value: 0. Expected lower limit: 0. Expected upper limit: 41.0496. A: 2e+018. B: 4e+020. C:-2.40674e+007.<br/>

This probably means it was rounded by BYOND to be zero, which means I'll need make the check >= rather than >. Simple fix really.


Inf for a,b, and c only started to appear on e17:
Starting at i = e17
Gas pressure pumping test finished. i=1e+016 Expected pressure = 2936.08 +/- 5. Got 2436.08.
[16:43:03] Runtime in stack_trace.dm,3: Failed to solve pressure quadratic equation. A: 1.#INF. B: 1.#INF. C:-2.40674e+007. Current value = -1.#IND

I expect the newton's method to truly fail here, but i'll retest in one moment.

**Comment by LemonInTheDark** [comment]
?

**Comment by vincentiusvin** [comment]
A bit concerned about very low gas transfers stack tracing. Thinking about actually changing it to <= but then I'll need to be sure that divisions by infinity or other inf memes never result in zero
Also some problems with transferring to an empty gasmix that I'll need to iron out. 
Probably can get back to this in a week or so?

**Comment by LemonInTheDark** [comment]
Sounds good to me

**Comment by LemonInTheDark** [comment]
Jesus christ

**Comment by vincentiusvin** [comment]
image

**Comment by LemonInTheDark** [comment]
Ready I take it? or are you still hung on that error

**Comment by vincentiusvin** [comment]
ill just keep the > and comment out the stack trace. Future contributors can always enable it if currently unforeseen bugs arise.

**Comment by vincentiusvin** [comment]
Stalking coderbus it seems that isinf isnan etc is planned for the future, that might be worth looking into when it finally comes.

**Comment by vincentiusvin** [comment]
Ready!

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
Wew, heck of a read

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 2]
```diff
@@ -513,70 +513,168 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
...
+	var/lower_limit = max((pv / rt_low) - output_air.total_moles(), 0)
+	var/upper_limit = (pv / rt_high) - output_air.total_moles() // In theory this should never go below zero, the pressure_delta check above should account for this.
+
+	/*
+	 * We have PV=nRT as a nice formula, we can rearrange it into nT = PV/R
+	 * But now both n and T can change, since any incoming moles also change our temperature.
+	 * So we need to unify both our n and T, somehow.
+	 * We can rewrite T as (our old thermal energy + incoming thermal energy) divided by (our old heat capacity + incoming heat capacity)
+	 * T = (W1 + n/N2 * W2) / (C1+ n/N2 *C2). C being heat capacity, W being work, N being total moles.
+	 * In total we now have our equation be: (N1 + n) * (W1 + n/N2 * W2) / (C1+ n/N2 * C2) = PV/R
```
[the remark this answers is not in the document]
Comment (vincentiusvin) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm:587]: Target's total moles.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 3]
```diff
@@ -513,70 +513,168 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
...
+	 * Now you can rearrange this and find out that it's a quadratic equation and pretty much solvable with the formula. Will be a bit messy though.
+	 * (W2/N2)n^2 + (N1*W2/N2)n + W1n - ((PV/R)*C2/N2)n - (PV/R)*C1 + N1W1 = 0
+	 * We will represent each of these terms with A, B, and C. A for the n^2 part, B for the n^1 part, and c for the n^0 part.
+	 * We then put this into the famous (-b +/- sqrt(b^2-4ac)) / 2a formula.
+	 * Oh, and one more thing. By "our" we mean the gasmix in the argument. We are the incoming one here. We are number 2, target is number 1.
+	 * If all this counting fucks up, we revert first to Newton's approximation, then the old simple formula.
+	 */
+
+	// The y in our equation.
```
[the remark this answers is not in the document]
Comment (vincentiusvin) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm:596]: yeah its a bit unclear, i'll change it

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 4]
```diff
@@ -513,70 +513,168 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
...
+	var/ntvalue = pv / R_IDEAL_GAS_EQUATION
+	// x^2
+	var/a_value = thermal_energy() / total_moles()
+	// x^1
+	var/b_value = (output_air.total_moles() * thermal_energy() / total_moles())  + output_air.thermal_energy() - (ntvalue*heat_capacity()/total_moles())
+	// x^0
+	var/c_value = (-1 * ntvalue * output_air.heat_capacity()) + (output_air.total_moles() * output_air.thermal_energy())
```
[the remark this answers is not in the document]
Comment (vincentiusvin) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm:603]: sure ill try that

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 5]
```diff
@@ -513,70 +513,168 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
...
 
-10/20*5 = 2.5
-10 = 2.5/5*20
-**/
+/// Approximation of the quadratic equation using Newton-Raphson's Method.
+/// We use the slope of an approximate value to get closer to the root of a given equation.
+/datum/gas_mixture/proc/gas_pressure_approximate(a, b, c, lower_limit, upper_limit)
+	if(!IS_INF_OR_NAN(a) && !IS_INF_OR_NAN(b) && !IS_INF_OR_NAN(c))
+		// We need to start off at a reasonably good estimate. For very big numbers the amount of moles is most likely small so better start with lower_limit.
+		. = lower_limit
```
[the remark this answers is not in the document]
Comment (vincentiusvin) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm:629]: its the returned value. that's the only reason.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 6]
```diff
@@ -513,70 +513,168 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
...
+	. = gas_pressure_approximate(a_value, b_value, c_value, lower_limit, upper_limit)
+	if(.)
+		return
+	// Inaccurate and will probably explode but whatever.
+	return (pressure_delta*output_air.volume)/(temperature * R_IDEAL_GAS_EQUATION)
+
+/// Actually tries to solve the quadratic equation.
+/// Do mind that the numbers can get very big and might hit BYOND's single point float limit.
+/datum/gas_mixture/proc/gas_pressure_quadratic(a, b, c, lower_limit, upper_limit)
+	if(!IS_INF_OR_NAN(a) && !IS_INF_OR_NAN(b) && !IS_INF_OR_NAN(c))
```
[the remark this answers is not in the document]
Comment (vincentiusvin) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm:617]: would mean writing the stack trace twice though.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 4 above)
[the remark this answers is not in the document]
Comment (vincentiusvin) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: f432553aa67f3b47e6ae581a004f86924d5bd97e

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 6 above)
[the remark this answers is not in the document]
Comment (vincentiusvin) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: 7f943d74f9c885729e34449efa2f300627457ffa

**Comment by vincentiusvin** [comment]
Oh seems like i botched something, on the bright side the tests work.

**Comment by vincentiusvin** [comment]
Master seems to be failing integration.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 7]
```diff
@@ -534,70 +534,190 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
...
+	if(IS_INF_OR_NAN(a) || IS_INF_OR_NAN(b) || IS_INF_OR_NAN(c))
+		stack_trace("Newton's Approximation for pressure failed, inf-nan check failed. A: [a]. B: [b]. C:[c]. Expected lower limit: [lower_limit]. Expected upper limit: [upper_limit]. ")
+		return FALSE
+	// We need to start off at a reasonably good estimate. For very big numbers the amount of moles is most likely small so better start with lower_limit.
+	var/solution = lower_limit
+	for (var/iteration in 1 to ATMOS_PRESSURE_APPROXIMATION_ITERATIONS)
+		var/diff = (a*solution**2 + b*solution + c) / (2*a*solution + b) // f(solution) / f'(solution)
+		solution -= diff // xn+1 = xn - f(solution) / f'(solution)
+		if(abs(diff) < MOLAR_ACCURACY && (solution > lower_limit) && (solution < upper_limit))
+			return
```
Comment (LemonInTheDark) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm:677]: don't you need to return the solution?

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 8]
```diff
@@ -534,70 +534,190 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
...
-			return TRUE
-		var/datum/gas_mixture/removed = remove(transfer_moles)
-		output_air.merge(removed)
-		return TRUE
+/// Approximation of the quadratic equation using Newton-Raphson's Method.
+/// We use the slope of an approximate value to get closer to the root of a given equation.
+/datum/gas_mixture/proc/gas_pressure_approximate(a, b, c, lower_limit, upper_limit)
+	if(IS_INF_OR_NAN(a) || IS_INF_OR_NAN(b) || IS_INF_OR_NAN(c))
+		stack_trace("Newton's Approximation for pressure failed, inf-nan check failed. A: [a]. B: [b]. C:[c]. Expected lower limit: [lower_limit]. Expected upper limit: [upper_limit]. ")
+		return FALSE
```
Comment (LemonInTheDark) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm:670]: I hadn't considered this properly. If you'd like you can keep the nesting

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 7 above)
Comment (vincentiusvin) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm:677]: woops

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 8 above)
Comment (vincentiusvin) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: 👍

**Pull request merged**